flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,190 @@
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.enableCheckpointing(10)
+                .enableChangelogStateBackend(true)
+                .getCheckpointConfig()
+                .setCheckpointStorage(TEMPORARY_FOLDER.newFolder().toURI());
+        env.setStateBackend(
+                        new DelegatedStateBackendWrapper(
+                                delegatedStateBackend,
+                                snapshotResultFuture -> {
+                                    RunnableFuture<SnapshotResult<KeyedStateHandle>>
+                                            snapshotResultFutureWrapper =
+                                                    new FutureTaskWithException<>(
+                                                            () -> {
+                                                                SnapshotResult<KeyedStateHandle>
+                                                                        snapshotResult =
+                                                                                snapshotResultFuture
+                                                                                        .get();
+                                                                if (PENDING_MATERIALIZATION.size()
+                                                                        == 0) {
+                                                                    throw new RuntimeException();
+                                                                } else {
+                                                                    return snapshotResult;
+                                                                }
+                                                            });
+                                    PENDING_MATERIALIZATION.add(snapshotResultFutureWrapper);
+                                    return snapshotResultFutureWrapper;
+                                }))
+                .setRestartStrategy(RestartStrategies.fixedDelayRestart(1, 0));
+        env.configure(
+                new Configuration()
+                        .set(
+                                StateChangelogOptions.PERIODIC_MATERIALIZATION_INTERVAL,
+                                Duration.ofMillis(20))
+                        .set(StateChangelogOptions.MATERIALIZATION_MAX_FAILURES_ALLOWED, 1));
+        waitAndAssert(
+                env,
+                buildStreamGraph(
+                        env,
+                        new ControlledSource() {
+                            @Override
+                            protected void beforeElement() throws Exception {
+                                waitWhile(
+                                        () ->
+                                                currentIndex >= TOTAL_ELEMENTS / 2
+                                                        && PENDING_MATERIALIZATION.size() == 0);
```

Comment:
Sure. After half of elements have been consumed, if the materialization is still not done (it will throw a exception and cause the materation process fail), just wait.

Reply:
Thanks for clarifying. Shouldn't this check also include `isDone` part?

 (or whatever it will be replaced with as per comments above)